### PR TITLE
POWER8 VSX Optimization

### LINKED ISSUE #397: Optimize dlib for POWER8 VSX
Enable and optimize support for POWER8 VSX SIMD instructions on PPC64LE Linux to dlib/simd.

$$$$ Financial bounties available.  Any reasonable suggested value will be seriously considered.

I welcome contact / replies from developers in the dlib community who are interested to work on this project.

### PR DESCRIPTION
**Pull request by barkovv**
@davisking 
Resolves #397 
Code can be compiled with gcc or clang in similar to SSE way:
```
cd dlib/test
mkdir build
cd build
cmake ..  -DUSE_VSX_INSTRUCTIONS=1
cmake --build . --config Release
./dtest --runall
```

### Current problems:
- [x] "hash" test fails on either optimized or not optimized version (but for x86 everything is ok). I will fix it.
- [x] I don't know how measure profit of optimisation. Is  there some profiler tool or something else for that purpose in dlib? Any help is appreciated

### CODE REVIEW

File: dlib/simd/simd_vsx.h [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#ifndef DLIB_sIMD_VSX_Hh_
+#define DLIB_sIMD_VSX_Hh_
+
+#include <altivec.h> // VSX
+#include <stdlib.h>
+
+/* Returns amount of bytes till previous 16-byte aligned value  */
+inline size_t getAlignOffset(const void *p) const {
+    return ((size_t) p) & 0x0f;// This is optimised variant of ((size_t) p) % 16
```
Comment (fenollp) [inline dlib/simd/simd_vsx.h]: Or you would think: https://godbolt.org/g/0XxEdU

**Comment by davisking** [comment]
The main pieces of code that use the simd objects are the extract_fhog_features(), resize_image(), and spatially_filter_image() routines.  These are the building blocks of the fhog object detection code, which is probably the main thing @edelsohn is interested in.  You can test the speed of the fhog detector by running the fhog face detector from this example program: http://dlib.net/face_detection_ex.cpp.html

**Comment by davisking** [comment]
I just merged https://github.com/davisking/dlib/pull/414 which resolves the same issue, so I'm closing this PR.  Feel free to reopen if you have additional changes you want to submit though, and thanks for submitting the PR in any case :)

**Comment by edelsohn** [comment]
@barkovv There are plenty of other bounty projects on POWER, if you're interested.

**Pull request closed without merging**